application/src/main/java/org/thingsboard/server/service/ttl/KafkaEdgeTopicsCleanUpService.java

```diff
@@ -0,0 +1,111 @@
+
+import java.time.Instant;
+import java.util.Date;
+import java.util.Optional;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+@Slf4j
+@Service
+@TbCoreComponent
+@RequiredArgsConstructor
+@ConditionalOnExpression("'${queue.type:null}'=='kafka' && ${edges.enabled:true}")
+public class KafkaEdgeTopicsCleanUpService {
+
+    private static final long ONE_MONTH_MILLIS = TimeUnit.DAYS.toChronoUnit().getDuration().multipliedBy(30).toMillis();
+
+    private final EdgeService edgeService;
+    private final TenantService tenantService;
+    private final AttributesService attributesService;
+
+    private final TopicService topicService;
+    private final PartitionService partitionService;
+
+    private final TbKafkaSettings kafkaSettings;
+    private final TbKafkaTopicConfigs kafkaTopicConfigs;
+
+    private final ExecutorService executorService = Executors.newSingleThreadExecutor(ThingsBoardThreadFactory.forName("kafka-edge-topic-cleanup"));
+
+    @Scheduled(initialDelayString = "#{T(org.apache.commons.lang3.RandomUtils).nextLong(0, ${sql.ttl.edge_events.execution_interval_ms})}", fixedDelayString = "${sql.ttl.edge_events.execution_interval_ms}")
+    public void cleanUp() {
+        executorService.submit(() -> {
+            PageDataIterable<TenantId> tenants = new PageDataIterable<>(tenantService::findTenantsIds, 10_000);
+            for (TenantId tenantId : tenants) {
+                try {
+                    cleanUp(tenantId);
+                } catch (Exception e) {
+                    log.warn("Failed to drop kafka topics for tenant {}", tenantId, e);
+                }
+            }
+        });
+    }
+
+    private void cleanUp(TenantId tenantId) throws Exception {
+        if (!partitionService.resolve(ServiceType.TB_CORE, tenantId, tenantId).isMyPartition()) {
+            return;
+        }
+
+        PageDataIterable<EdgeId> edgeIds = new PageDataIterable<>(link -> edgeService.findEdgeIdsByTenantId(tenantId, link), 1024);
+        long currentTimeMillis = System.currentTimeMillis();
+
+        for (EdgeId edgeId : edgeIds) {
+            Optional<AttributeKvEntry> attributeOpt = attributesService.find(tenantId, edgeId, AttributeScope.SERVER_SCOPE, DefaultDeviceStateService.LAST_CONNECT_TIME).get();
+            if (attributeOpt.isPresent()) {
+                Optional<Long> lastConnectTimeOpt = attributeOpt.get().getLongValue();
+                if (lastConnectTimeOpt.isPresent() && isTopicExpired(lastConnectTimeOpt.get(), currentTimeMillis)) {
+                    String topic = topicService.buildEdgeEventNotificationsTopicPartitionInfo(tenantId, edgeId).getTopic();
+                    TbKafkaAdmin kafkaAdmin = new TbKafkaAdmin(kafkaSettings, kafkaTopicConfigs.getEdgeEventConfigs());
+                    if (kafkaAdmin.isTopicEmpty(topic)) {
+                        kafkaAdmin.deleteTopic(topic);
```

Comment:
What about the consumer that might be running at this time? Or it is guaranteed that it won't?

Reply:
Yeap, if Edge was connected last time more than 1 month ago